gtsam/discrete/TableFactor.h

```diff
@@ -0,0 +1,195 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file TableFactor.h
+ * @date May 12, 2022
+ * @author Yoonwoo Kim
+ */
+
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/discrete/DecisionTreeFactor.h>
+#include <gtsam/discrete/DiscreteFactor.h>
+#include <gtsam/discrete/DiscreteKey.h>
+#include <gtsam/inference/Ordering.h>
+
+#include <Eigen/Sparse>
+#include <algorithm>
+#include <boost/assign/std/map.hpp>
+#include <boost/make_shared.hpp>
+#include <boost/shared_ptr.hpp>
+#include <cmath>  // for floor
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace gtsam {
+
+class GTSAM_EXPORT TableFactor : public DiscreteFactor {
+ private:
+  Eigen::SparseVector<double> sparse_table_;
+  std::map<Key, size_t> cardinalities_;
+  std::map<Key, size_t> denominator_;
+
+  DiscreteKey discreteKey(size_t i) const {
+    Key j = keys_[i];
+    return DiscreteKey(j, cardinalities_.at(j));
+  }
+
+  /** Convert table to SparseVector */
+  static Eigen::SparseVector<double> Convert(const std::vector<double>& table);
+
+  /** Convert table to SparseVector */
+  static Eigen::SparseVector<double> Convert(const std::string& table);
+
+ public:
+  // typedefs needed to play nice with gtsam
+  typedef TableFactor This;
+  typedef DiscreteFactor Base;  ///< Typedef to base class
+  typedef boost::shared_ptr<TableFactor> shared_ptr;
+
+  /// @name Constructors
+  /// @{
+
+  /** Default constructo for I/O */
+  TableFactor();
+
+  /** Constructor from Eigen::SparseVector */
+  TableFactor(const DiscreteKeys& keys,
+              const Eigen::SparseVector<double>& table);
+
+  /** Constructor from doubles */
+  TableFactor(const DiscreteKeys& keys, const std::vector<double>& table)
+      : TableFactor(keys, Convert(table)) {}
+
+  /** Constructor from string */
+  TableFactor(const DiscreteKeys& keys, const std::string& table)
+      : TableFactor(keys, Convert(table)) {}
+
+  /// Single-key specialization
```

Comment:
This constructor is written to construct a factor for one variable (DiscreteKey instead of DiscreteKeys). Should I rephrase to "Constructor for single-key"?